Add support for modmap None.

wismill opened the pull request:
Fixes #211

wismill commented:
@bluetech could you guide me to write the corresponding tests?

wismill commented:
@bluetech @whot kind reminder for a review.

whot commented:
ETIME on the actual code but please add the information from #211 to the commit message so it's not 3 clicks away from anyone trying to figure out why this was added.

whot commented on src/darray.h:
```diff
@@ -143,6 +143,13 @@ typedef darray (unsigned long)  darray_ulong;
 
+#define darray_remove(arr, i) do { \
+	size_t index_ = (i); \
+	if (index_ < arr.size-1)    \
```
style nitpick:
```suggestion
	if (index_ < arr.size-1) \
```

also, `arr.size` and `size_t` are both unsigned so removing any index from a zero-sized array will proceed without a warning (or rather: it's not a noop as expected). Might be better to use `ssize_t` here to check for `<0` as well.

I strongly recommend splitting the `darray_remove` addition into a separate patch and adding test cases for it so we don't have to worry about this. There's nothing existing that I can find, but a `test/darray.c` won't hurt :)

wismill replied:
Done

whot commented on src/darray.h:
```diff
@@ -143,6 +143,13 @@ typedef darray (unsigned long)  darray_ulong;
+#define darray_remove(arr, i) do { \
+	size_t index_ = (i); \
+	if (index_ < arr.size-1)    \
+		memmove(&(arr).item[index_], &(arr).item[index_+1], ((arr).size-1-index_)*sizeof(*(arr).item)); \
```
style nitpick
```suggestion
		memmove(&(arr).item[index_], &(arr).item[index_+1], ((arr).size-1 - index_)*sizeof(*(arr).item)); \
```

wismill replied:
Done

whot commented on src/xkbcomp/symbols.c:
```diff
@@ -462,6 +463,12 @@ AddModMapEntry(SymbolsInfo *info, ModMapEntry *new)
         use = (clobber ? new->modifier : old->modifier);
         ignore = (clobber ? old->modifier : new->modifier);
 
+        // Remove entry
```
This file only uses `/* */`-style comments

wismill replied:
Done

whot reviewed (changes requested):
A few style fixes and the darray bit needs fixing, but other than that this LGTM (though I'm not well-versed in the internals of the actual modifier mappings)

whot reviewed (approved):
thanks, this LGTM

wismill commented:
> thanks, this LGTM

@whot I think we should add a test for the new feature. Can you give me  some guidance?

wismill commented:
Kind reminder

wismill commented:
@whot @bluetech What is blocking this PR? Kind reminder.

bluetech reviewed (commented):
I've read the xkbcomp code, and after much sighing over the wonders of 90s style C code, and after finally figuring out that the entire thing relies on the fact that `1U << entry->modifier` ends up being zero when `entry->modifier` is `XkbNoModifier` (i.e. 0xFF), I think I understand it.

So here are my review comments:

1. No reason to expose `XKB_NO_MODIFIER` in the public header.

3. In fact, will be better not to use `XKB_NO_MODIFIER` at all, since it's in-band data that is prone to error. Instead, should just add a flag to `ModMapEntry` saying "this is no-modifier/clearing entry". Luckily this stuff is entirely encapsulated in xkbcommon so it's not a problem.

5. The mechanism by which `None` works in xkbcomp is that the `None` entry overrides conflicting entries using the usual merge mechanism, and ends up being a no-op in `CopyModMapDefToKeymap`. This PR instead makes `None` physically remove conflicting entries (and not adding a "no-modifier" entry). I think the xkbcomp approach is preferable, because:
   1. It preserves compat with xkbcomp.
   2. It preserve the usual merging semantics which is what is expected.
   3. It's simpler (no need for `darray_remove` which is better avoided in general).

WDYT?

wismill commented:
@bluetech Thanks for the code archaeology! Now I see how to implement it.

I followed your suggestion and added an extensive comment in the commit to explain the behaviour in X11’s xkbcomp and why we do not use their trick. Please review that I got the correct understanding!

I reverted the previous commits, rebased and added some tests.

bluetech commented on src/xkbcomp/symbols.c:
```diff
@@ -1152,14 +1152,21 @@ HandleModMapDef(SymbolsInfo *info, ModMapDef *def)
+
+    if (def->modifier == none) {
+        // Handle special "None" entry
+        ndx = XKB_MOD_NONE;
```
Let's add a comment on the ModMapDef `modifier` field in the struct definition, `// Can also be XKB_MOD_NONE, meaning "don't add a modifier to the modmap".`

wismill replied:
I think you meant ModMapEntry. I added a comment to both.

bluetech reviewed (commented):
Thanks for the update, LGTM with just a couple of small comments.

wismill commented on test/data/keymaps/modmap-none.xkb:
```diff
@@ -0,0 +1,93 @@
+    modmap Shift { Shift_R };
+    modmap None  { Shift_R };
+
+    // Cycle
```
I added some. Do you have other in mind?

bluetech replied:
No, the test is great.

wismill commented on src/xkbcomp/symbols.c:
```diff
@@ -1152,14 +1152,21 @@ HandleModMapDef(SymbolsInfo *info, ModMapDef *def)
-        return false;
+    xkb_atom_t none = xkb_atom_intern(ctx, "None", 4);
+
+    if (def->modifier == none) {
```
Fixed

wismill commented on src/keymap.h:
```diff
@@ -109,6 +109,9 @@
 #define XKB_MAX_LEDS ((xkb_led_index_t) (sizeof(xkb_led_mask_t) * 8))
 
+/* Special value to handle modMap None {…} */
+#define XKB_MOD_NONE 0xffffffff
```
Fixed
